File: py/queue.py
from collections import deque

class Queue:
    def __init__(self):
        self.items = deque()
    
    def add(self, data):
        self.items.appendleft(data)

    def pop(self):
        try:
            self.items.pop()
        except IndexError:
            raise RuntimeError('pop from empty Queue')

    def __len__(self):
        return len(self.items)




q = Queue()
q.add(3)
assert(len(q) == 1)
q.add(2)
q.pop()
q.pop()

File: js/queue.js
// inefficient, but quick and dirty
// unshift is an O(n) operation

class Queue {
    constructor() {
        this.items = [];
    }
    add(data) {
        this.items.unshift(data);
    }
    pop() {
        this.items.pop();
    }
    top() {
        return this.items[this.items.length-1];
    }
}

s = new Queue();
s.add(3);
s.add(5);
s.pop();
s.add(7);
console.log(s.top());
s.add(12);